src/sunadaptcontroller/soderlind/sunadaptcontroller_soderlind.c

```diff
@@ -0,0 +1,393 @@
+#define DEFAULT_PID_K3      RCONST(0.1)
+#define DEFAULT_PI_K1       RCONST(0.8)     /* PI parameters */
+#define DEFAULT_PI_K2      -RCONST(0.31)
+#define DEFAULT_I_K1        RCONST(1.0)     /* I parameters */
+#define DEFAULT_EXPGUS_K1   RCONST(0.367)   /* Explicit Gustafsson parameters */
+#define DEFAULT_EXPGUS_K2   RCONST(0.268)
+#define DEFAULT_IMPGUS_K1   RCONST(0.98)    /* Implicit Gustafsson parameters */
+#define DEFAULT_IMPGUS_K2   RCONST(0.95)
+#define DEFAULT_BIAS        RCONST(1.5)
+#define TINY                RCONST(1.0e-10)
+
+
+/* -----------------------------------------------------------------
+ * exported functions
+ * ----------------------------------------------------------------- */
+
+/* -----------------------------------------------------------------
+ * Function to create a new Soderlind controller
+ */
+
+SUNAdaptController SUNAdaptController_Soderlind(SUNContext sunctx)
+{
+  SUNAdaptController C;
+  SUNAdaptControllerContent_Soderlind content;
+
+  /* Create an empty controller object */
+  C = NULL;
+  C = SUNAdaptController_NewEmpty(sunctx);
+  if (C == NULL) { return (NULL); }
+
+  /* Attach operations */
+  C->ops->gettype      = SUNAdaptController_GetType_Soderlind;
+  C->ops->estimatestep = SUNAdaptController_EstimateStep_Soderlind;
+  C->ops->reset        = SUNAdaptController_Reset_Soderlind;
+  C->ops->setdefaults  = SUNAdaptController_SetDefaults_Soderlind;
+  C->ops->write        = SUNAdaptController_Write_Soderlind;
+  C->ops->seterrorbias = SUNAdaptController_SetErrorBias_Soderlind;
+  C->ops->updateh      = SUNAdaptController_UpdateH_Soderlind;
+  C->ops->space        = SUNAdaptController_Space_Soderlind;
+
+  /* Create content */
+  content = NULL;
+  content = (SUNAdaptControllerContent_Soderlind)malloc(sizeof *content);
+  if (content == NULL)
+  {
+    (void) SUNAdaptController_Destroy(C);
+    return (NULL);
+  }
+
+  /* Attach content */
+  C->content = content;
+
+  /* Fill content with default/reset values */
+  SUNAdaptController_SetDefaults_Soderlind(C);
+  SUNAdaptController_Reset_Soderlind(C);
+
+  return (C);
+}
+
+/* -----------------------------------------------------------------
+ * Function to set Soderlind parameters
+ */
+
+int SUNAdaptController_SetParams_Soderlind(SUNAdaptController C,
+                                           sunrealtype k1, sunrealtype k2,
+                                           sunrealtype k3, sunrealtype k4,
+                                           sunrealtype k5)
+{
+  SODERLIND_K1(C) = k1;
+  SODERLIND_K2(C) = k2;
+  SODERLIND_K3(C) = k3;
+  SODERLIND_K4(C) = k4;
+  SODERLIND_K5(C) = k5;
+  return SUNADAPTCONTROLLER_SUCCESS;
+}
+
+/* -----------------------------------------------------------------
+ * Function to create a PID controller (subset of Soderlind)
+ */
+
+SUNAdaptController SUNAdaptController_PID(SUNContext sunctx)
+{
+  SUNAdaptController C;
+  C = SUNAdaptController_Soderlind(sunctx);
+  (void) SUNAdaptController_SetParams_PID(C,
```

Comment:
It unlikely but we should check this return value, destroy the controller, and return `NULL` if setting the parameters (somehow) fails.

Reply:
See comments below about checking for a `NULL` input